### PR TITLE
Classification queue

### PR DESCRIPTION
**Pull request by marten**
Extracts a ClassificationQueue, making the classify page a little lighter on the logic. Also means that this process is now more easily tested.

Staging branch URL: https://classification-queue.pfe-preview.zooniverse.org

Fixes # .

Describe your changes.

# Required Manual Testing

- [ ] Does the non-logged in home page render correctly?
- [ ] Does the logged in home page render correctly?
- [ ] Does the projects page render correctly?
- [ ] Can you load project home pages?
- [ ] Can you load the classification page?
- [ ] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

### CODE REVIEW

File: app/lib/classification-queue.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+        this.recents.unshift(classification);
+    }
+
+    flushToBackend() {
+        var queue = this._loadQueue();
+
+        if (queue.length > 0) {
+            console.log('Saving queued classifications:', queue.length);
+            for (let classificationData of queue) {
```
Comment (eatyourgreens) [inline app/lib/classification-queue.js]: `for…of` isn’t supported by arrays in IE. I think we’ve preferred `Array.forEach` for better browser compatibility so `queue.forEach( function(classificationData) {…})` might be safer here.

File: app/pages/project/classify.cjsx [hunk 2]
```diff
@@ -9,15 +9,14 @@ counterpart = require 'counterpart'
 `import FinishedBanner from './finished-banner';`
 Classifier = require('../../classifier').default
 seenThisSession = require '../../lib/seen-this-session'
+ClassificationQueue = require('../../lib/classification-queue').default
 `import WorkflowAssignmentDialog from '../../components/workflow-assignment-dialog';`
 { Split } = require('seven-ten')
```
Comment (eatyourgreens) [inline app/pages/project/classify.cjsx]: Is Split still needed here?

File: app/lib/classification-queue.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                this.apiClient.type('classifications').create(classificationData)
+                    .save()
+                    .then((actualClassification) => {
+                        console.log('Saved classification', actualClassification.id);
+                        Split.classificationCreated(actualClassification); // Metric log needs classification id
```
Comment (eatyourgreens) [inline app/lib/classification-queue.js]: We should find out if Split is still needed here.

Comment (eatyourgreens) [reply app/lib/classification-queue.js]: Actually, would it be better to pass a post-classification callback to the constructor, invoke it here and leave `Split` in the calling page?

File: app/lib/classification-queue.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                            try {
+                                this._saveQueue(queue);
+                            }
+                            catch(error) {
+                                console.error('Failed to update classification queue:', error);
+                            }
+                            break;
+                        }
+                    });
+            }
```
Comment (eatyourgreens) [inline app/lib/classification-queue.js]: Indentation is weird here. eslint should be complaining about this.

**Comment by eatyourgreens** [comment]
Tests are warning about an unhandled Promise rejection, so there's a catch block missing somewhere.

File: app/lib/classification-queue.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            for (let classificationData of queue) {
+                this.apiClient.type('classifications').create(classificationData)
+                    .save()
+                    .then((actualClassification) => {
+                        console.log('Saved classification', actualClassification.id);
+                        Split.classificationCreated(actualClassification); // Metric log needs classification id
+                        this.addRecent(actualClassification);
+                        let indexInQueue = queue.indexOf(classificationData);
+                        queue.splice(indexInQueue, 1);
+                        try {
```
Comment (eatyourgreens) [inline app/lib/classification-queue.js]: I'm not sure why this is using `try…catch` instead of handling errors when the promise rejects.

Comment (eatyourgreens) [reply app/lib/classification-queue.js]: nvm. This is catching local storage errors, like 'out of space'.

File: test/fake-api-client.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    save() {
+        let saved = this.resourceType.save(this.attributes);
+
+        if (saved) {
+            return Promise.resolve(this);
+        }
+        else {
+            return new Promise((resolve, reject) => { reject(); });
```
Comment (eatyourgreens) [inline test/fake-api-client.js]: `reject( new Error('Save failed!') )` would be better here. The tests currently log `Failed to save a queued classification: undefined`.

File: app/lib/classification-queue.js (hunk 1 above)
Comment (eatyourgreens) [reply app/lib/classification-queue.js]: Seems to work in IE11 on BrowserStack, so I might be too paranoid about browser compatibility here.

**Comment by eatyourgreens** [comment]
https://classification-queue.pfe-preview.zooniverse.org/dev/classifier throws `Uncaught TypeError: Cannot read property 'catch' of undefined` when you finish a classification. I think that's an existing bug so not a blocking issue here.

File: app/lib/classification-queue.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+class ClassificationQueue {
+    constructor(storage, api) {
+        this.storage = storage || window.localStorage;
+        this.apiClient = api || apiClient;
+        this.recents = [];
+    }
+
+    add(classification) {
+        var queue = this._loadQueue();
```
Comment (eatyourgreens) [inline app/lib/classification-queue.js]: Should be `const queue`.

File: app/lib/classification-queue.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        }
+        catch(error) {
+            console.error('Failed to queue classification:', error);
+        }
+
+        this.flushToBackend();
+    }
+
+    length() {
+        return this._loadQueue().length
```
Comment (eatyourgreens) [inline app/lib/classification-queue.js]: Missing semicolon.

File: app/lib/classification-queue.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        if (this.recents.length > MAX_RECENTS) {
+            let droppedClassification = this.recents.pop();
+            droppedClassification.destroy();
+        }
+
+        this.recents.unshift(classification);
+    }
+
+    flushToBackend() {
+        var queue = this._loadQueue();
```
Comment (eatyourgreens) [inline app/lib/classification-queue.js]: `const queue`

File: app/lib/classification-queue.js (hunk 1 above)
Comment (eatyourgreens) [reply app/lib/classification-queue.js]: `const` instead of `let` here.

File: app/lib/classification-queue.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        if (queue.length > 0) {
+            console.log('Saving queued classifications:', queue.length);
+            for (let classificationData of queue) {
+                this.apiClient.type('classifications').create(classificationData)
+                    .save()
+                    .then((actualClassification) => {
+                        console.log('Saved classification', actualClassification.id);
+                        Split.classificationCreated(actualClassification); // Metric log needs classification id
+                        this.addRecent(actualClassification);
+                        let indexInQueue = queue.indexOf(classificationData);
```
Comment (eatyourgreens) [inline app/lib/classification-queue.js]: `const indexInQueue`

File: app/lib/classification-queue.js [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+                            console.error('Failed to update classification queue:', error);
+                        }
+                    })
+                    .catch(function(error) {
+                        console.error('Failed to save a queued classification:', error);
+
+                        switch(error.status) {
+                        case 422:
+                            console.error("Dropping malformed classification permanently", classificationData);
+                            let indexInQueue = queue.indexOf(classificationData);
```
Comment (eatyourgreens) [inline app/lib/classification-queue.js]: `const indexInQueue`

File: app/lib/classification-queue.js [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+                            }
+                            break;
+                        }
+                    });
+            }
+        }
+    }
+
+    _loadQueue() {
+        var queue = JSON.parse(this.storage.getItem(FAILED_CLASSIFICATION_QUEUE_NAME));
```
Comment (eatyourgreens) [inline app/lib/classification-queue.js]: `const queue`

**Comment by eatyourgreens** [comment]
eslint throws a ton of warnings, mostly about unexpected indentation throughout. I've added comments for some of the syntax changes it suggests.

File: app/lib/classification-queue.js (hunk 9 above)
Comment (marten) [reply app/lib/classification-queue.js]: But as it's sending stuff out, it's removing elements from the array. I know that we're not changing the "pointer", but marking it const seems like a lie to readers of the code?

Comment (eatyourgreens) [reply app/lib/classification-queue.js]: It's never reassigned so it's `const` in JS.

File: app/lib/classification-queue.js (hunk 3 above)
Comment (srallen) [reply app/lib/classification-queue.js]: The metric logs Seven Ten makes needs the classification id if the split is tracking the classification created to find out if the split has an effect on classification. The researcher can then compare the split assigned to the user to the classification data export using the id.

File: app/lib/classification-queue.js (hunk 1 above)
Comment (marten) [reply app/lib/classification-queue.js]: There's a `babel-plugin-transform-es2015-for-of` in the package.lock. Might be that?

Comment (eatyourgreens) [reply app/lib/classification-queue.js]: Could well be.

File: app/lib/classification-queue.js (hunk 3 above)
Comment (srallen) [reply app/lib/classification-queue.js]: A callback would be fine for that and I think preferable long term.

**Comment by srallen** [comment]
This is approved, but labeled WIP still. Is the merge being delayed?

File: app/pages/project/classify.cjsx (hunk 2 above)
Comment (marten) [reply app/pages/project/classify.cjsx]: Yep

**Comment by marten** [comment]
@srallen I wanted to make sure to fix @eatyourgreens comments, but I think it's good to go now. Apologies for the incorrect indentation.

**Comment by eatyourgreens** [comment]
Looks good to me. Tests are passing and the queue is working with my test project on staging.

Both tests throw the 'Save failed!' error, which I wasn't expecting, but everything seems to be working. I'm happy to merge this.

**Pull request merged**